producer/src/blockchain/blockchain_test.go

```diff
@@ -240,3 +240,81 @@ func TestPhaseTwoMultiple(t *testing.T) {
 		t.Errorf("Blocks do not match (block 2 by height)")
 	}
 }
+
+func TestRecoverBlockchainMetadata(t *testing.T) {
+	blockchain := "testBlockchain.dat"
+	table := "table.db"
+	setUp(blockchain, table)
+	defer tearDown(blockchain, table)
+	// Create a bunch of blocks
+	block0 := block.Block{
+		Version:        1,
+		Height:         0,
+		Timestamp:      time.Now().UnixNano(),
+		PreviousHash:   block.HashSHA256([]byte{'0'}),
+		MerkleRootHash: block.HashSHA256([]byte{'1'}),
+		Data:           [][]byte{block.HashSHA256([]byte{'x', 'o', 'x', 'o'})},
+	}
+	block0.DataLen = uint16(len(block0.Data))
+	block1 := block.Block{
+		Version:        1,
+		Height:         1,
+		Timestamp:      time.Now().UnixNano(),
+		PreviousHash:   block.HashBlock(block0),
+		MerkleRootHash: block.HashSHA256([]byte{'1'}),
+		Data:           [][]byte{block.HashSHA256([]byte{'x', 'y', 'z'})},
+	}
+	block1.DataLen = uint16(len(block1.Data))
+	block2 := block.Block{
+		Version:        1,
+		Height:         2,
+		Timestamp:      time.Now().UnixNano(),
+		PreviousHash:   block.HashBlock(block1),
+		MerkleRootHash: block.HashSHA256([]byte{'1'}),
+		Data:           [][]byte{block.HashSHA256([]byte{'a', 'b', 'c'})},
+	}
+	block2.DataLen = uint16(len(block2.Data))
+	// Add all the blocks
+	err := AddBlock(block0, blockchain, table)
+	if err != nil {
+		t.Errorf("Failed to add block0.")
+	}
+	err = AddBlock(block1, blockchain, table)
+	if err != nil {
+		t.Errorf("Failed to add block1.")
+	}
+	err = AddBlock(block2, blockchain, table)
+	if err != nil {
+		t.Errorf("Failed to add block2.")
+	}
+	os.Remove(table)
```

Comment:
For some reason, upon conclusion of the test, we still have `table.db` remaining in our package. The `defer` should be removing it, and this line should also initially remove it. Can you put a check here to see if `table` exists, similar to how you implemented it in the function? And if it still exists, fail the test immediately.

Reply:
I put a check there right after` os.Remove(table)` and the test still passed. 

![image](https://user-images.githubusercontent.com/42977312/55272105-629b8c80-5285-11e9-86e3-32dd4f88f72e.png)

Reply:
I think `table.db` still exist after the unit test has ended is because there isn't a file removal call for `table.db` after the unit test called `RecoverBlockchainMetadata`

Reply:
Try editing the teardown function to check whether it's properly removing the table.

Reply:
The defer statement should remove it no matter what. You could try duplicating the defer statement after the Recover function call maybe.

Reply:
The `table.db` file is still there even after I put `os.Remove(table)` at the very end of the unit test.

Reply:
Check your function to make sure you're closing it properly. The os will not remove files still considered open.

Reply:
I figured out the issue. Apparently `os.Create()` creates the file and also opens the file.